File: problem5/dcp5.py
def cons(a, b):
    def pair(f):
        return f(a, b);
    return pair

def car(pair):
#    return pair(lambda x, y: x)
    def left(left, right):
        return left
    return pair(left)

def cdr(pair):
#    return pair(lambda x, y: y)
    def right(left, right):
        return right
    return pair(right)

print(car(cons(3, 4)))
print(cdr(cons(3, 4)))

File: problem5/dcp5.rb
def cons(a,b)
  pair = ->(f) {
    return f.call(a, b)
  }
  return pair
end

def car(f)
  left = ->(a, b) { return a }
  return f.call(left)
end

def cdr(f)
  right = ->(a, b) { return b }
  return f.call(right)
end

puts "left: #{car(cons(3, 4))}"
puts "right: #{cdr(cons(3, 4))}"
